sampo/userinput/parser/general_build.py

```diff
@@ -16,59 +17,140 @@
+        self.graph[u].append((v, weight))
+
+    def dfs_cycle(self, u, visited):
+        stack = [u]
+
+        while len(stack) > 0:
+            v = stack.pop()
+            if visited[v] == 1:
+                visited[v] = 2
+                continue
+            if visited[v] == 2:
+                continue
+            visited[v] = 1
+            # path.append(v)
+            stack.append(v)
+            # if len(self.graph[v]) == 0:
+            #     path = []
+            for neighbour, weight in self.graph[v]:
+                if visited[neighbour] == 0:
+                    stack.append(neighbour)
+                elif visited[neighbour] == 1:
+                    # path.append(neighbour)
+                    return True
+        return False
+
+    def find_cycle(self):
+        visited = {}
+        for key, value in self.graph.items():
+            for v in value:
+                visited[v[0]] = 0
+            visited[key] = 0
+        path = []
+
+        for key, value in visited.items():
+            if value == 0:
+                if self.dfs_cycle(key, visited):
+                    v = key
+                    path.append(v)
+                    while v not in path[:-1]:
+                        for n, _ in self.graph[v]:
+                            if visited[n] == 1:
+                                v = n
+                                break
+                        path.append(v)
+                    for v in path:
+                        visited[v] = 2
+                    return path
+
+        return None
+
+    def eliminate_cycle(self, cycle):
+        min_weight = float('inf')
+        min_edge = None
+
+        for i in range(len(cycle) - 1):
+            u = cycle[i]
+            v = cycle[i + 1]
+
+            for neighbor, weight in self.graph[u]:
+                try:
+                    if neighbor == v and weight < min_weight:
+                        min_weight = weight
+                        min_edge = (u, v)
+                except Exception:
+                    raise Exception(f'weight: {type(weight)} and min_weight: {type(min_weight)}\nv:{v} and u: {u}')
+
+        u, v = min_edge
+        self.graph[u] = [(neighbor, weight) for neighbor, weight in self.graph[u] if neighbor != v]
+
+    def eliminate_cycles(self, eliminate_cycle: bool = True):
+        cycle = self.find_cycle()
+        cycles = []
+
+        while cycle is not None:
+            for i in range(len(cycle)):
+                if cycle[i] == cycle[-1]:
+                    cycle = cycle[i:]
+                    break
+            if not eliminate_cycle:
+                cycles.append(cycle)
+            else:
+                self.eliminate_cycle(cycle)
+            cycle = self.find_cycle()
+        if not eliminate_cycle:
+            if not cycles:
+                return cycles
+            return None
+        for v in self.graph:
+            self.graph[v] = [u for u, w in self.graph[v]]
```

Comment:
Return statements are reachable not at all execution paths of function. Additionally, return type not specified.

Reply:
This method works in two modes: when we want to get all loops and the state of the graph after that is unimportant, when we want to exclude loops. For the first mode, the 'for' loop is not needed

Reply:
This should be in method documentation. 
And, please, specify the return type.